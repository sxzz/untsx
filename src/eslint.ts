import * as tsEsParser from '@typescript-eslint/parser'
import * as espree from 'espree'
import type { UntsxInstance } from '.'
import type { Linter } from 'eslint'

export function createESLintParser(
  espreeParse: UntsxInstance['espree'],
  eslintTypescriptParser: UntsxInstance['eslintTypescriptParser'],
): {
  jsParser: Linter.Parser
  tsParser: Linter.Parser
} {
  const jsParser: typeof espree = {
    ...espree,
    parse: espreeParse,
  }

  const tsParser: typeof tsEsParser = {
    ...tsEsParser,
    parseForESLint: eslintTypescriptParser,
  }
  ;(tsParser as any).parse = undefined

  return { jsParser, tsParser }
}
