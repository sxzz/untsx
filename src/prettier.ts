// @ts-expect-error
import { parsers as prettierEspree } from 'prettier/plugins/acorn.mjs'
// @ts-expect-error
import { parsers as prettierBabel } from 'prettier/plugins/babel.mjs'
// @ts-expect-error
import { printers } from 'prettier/plugins/estree.mjs'
// @ts-expect-error
import { parsers as prettierTS } from 'prettier/plugins/typescript.mjs'
import { REGEX_JSX, type UntsxFactory, type UntsxInstance } from '.'
import type { ParserOptions } from '@babel/parser'
import type { Comment } from 'acorn'
import type { Parser, Plugin } from 'prettier'

const babelOptions: ParserOptions = {
  sourceType: 'module',
  allowImportExportEverywhere: true,
  allowReturnOutsideFunction: true,
  allowNewTargetOutsideFunction: true,
  allowSuperOutsideMethod: true,
  allowUndeclaredExports: true,
  errorRecovery: true,
  createParenthesizedExpressions: true,
  createImportExpressions: true,
  attachComment: false,
  plugins: [
    'doExpressions',
    'exportDefaultFrom',
    'functionBind',
    'functionSent',
    'throwExpressions',
    'partialApplication',
    'decorators',
    'moduleBlocks',
    'asyncDoExpressions',
    'destructuringPrivate',
    'decoratorAutoAccessors',
    'sourcePhaseImports',
    'deferredImportEvaluation',
    ['optionalChainingAssign', { version: '2023-07' }],
  ],
  tokens: false,
  ranges: false,
}

export function createPrettierPlugin(
  { shouldTransform, isTarget, getVisitorKeys, format }: UntsxFactory,
  acorn: UntsxInstance['acorn'],
  babel: UntsxInstance['babel'],
  espree: UntsxInstance['espree'], // TODO
  eslintTypescript: UntsxInstance['eslintTypescript'],
): Plugin {
  const acornParser: Parser = {
    ...prettierEspree.acorn,
    parse(text, options) {
      if (shouldTransform && !shouldTransform(text, options.filepath)) {
        return prettierEspree.acorn.parse(text, options)
      }

      const comments: Comment[] = []
      const ast = acorn(text, {
        ecmaVersion: 'latest',
        sourceType: 'module',
        onComment: comments,
        allowReserved: true,
        allowReturnOutsideFunction: true,
        allowSuperOutsideMethod: true,
        allowImportExportEverywhere: true,
        checkPrivateFields: false,
        locations: false,
        ranges: true,
        preserveParens: true,
      })
      // @ts-expect-error -- expected
      ast.comments = comments
      return ast
    },
  }

  const babelParser: Parser = {
    ...prettierBabel.babel,
    parse(text, options) {
      if (shouldTransform && !shouldTransform(text, options.filepath)) {
        return prettierBabel.babel.parse(text, options)
      }

      return babel(text, babelOptions)
    },
  }
  const babelTsParser: Parser = {
    ...prettierBabel['babel-ts'],
    parse(text, options) {
      if (shouldTransform && !shouldTransform(text, options.filepath)) {
        return prettierBabel['babel-ts'].parse(text, options)
      }

      return babel(text, {
        ...babelOptions,
        plugins: [
          'typescript',
          ...(REGEX_JSX.test(options.filepath) ? ['jsx' as const] : []),
          ...(babelOptions.plugins || []),
        ],
      })
    },
  }

  const typescriptParser: Parser = {
    ...prettierTS.typescript,
    parse(text, options) {
      if (shouldTransform && !shouldTransform(text, options.filepath)) {
        return prettierTS.typescript.parse(text, options)
      }

      return eslintTypescript(text, {
        sourceType: 'module',
        loc: true,
        range: true,
        comment: true,
        tokens: false,
        loggerFn: false,
        project: false,
        jsDocParsingMode: 'none',
        suppressDeprecatedPropertyWarnings: true,
        filePath: options.filepath,
      })
    },
  }

  return {
    parsers: {
      acorn: acornParser,
      babel: babelParser,
      'babel-ts': babelTsParser,
      typescript: typescriptParser,
    },
    printers: {
      estree: {
        ...printers.estree,
        print(path, ...args) {
          if (isTarget(path.node)) {
            return format(path, ...args)
          }
          return printers.estree.print(path, ...args)
        },

        getVisitorKeys: getVisitorKeys
          ? (node, ...args) => {
              if (isTarget(node)) {
                return getVisitorKeys(node, ...args)
              }
              return printers.estree.getVisitorKeys(node, ...args)
            }
          : printers.estree.getVisitorKeys,
      },
    },
  }
}
