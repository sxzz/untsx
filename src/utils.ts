import { walk } from 'estree-walker'

export function getNodeRange(node: any): [start: number, end: number] {
  return node.range || [node.start, node.end]
}

export function patchLocOffset(node: any, offset?: number): any {
  if (offset === undefined) return node
  walk(node as any, {
    enter(node: any) {
      if (typeof node.loc === 'object') {
        node.loc.start.index += offset
        node.loc.end.index += offset
      }
      if (typeof node.range === 'object') {
        node.range[0] += offset
        node.range[1] += offset
      }
      if (typeof node.start === 'number') {
        node.start += offset
      }
      if (typeof node.end === 'number') {
        node.end += offset
      }
    },
  })
  return node
}
