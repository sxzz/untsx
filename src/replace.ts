import { walk } from 'estree-walker'
import { getNodeRange } from './utils'
import type { IsTargetFn } from '.'

export type NodeInfo = [
  code: string,
  start: number,
  end: number,
  validStart: number,
  validEnd: number,
]

export type ParseFn = (
  src: string,
  isExpression: boolean,
  offset?: number,
) => any
export type BuildFn = (start: number, end: number, valid: any) => any

export function replace<T>(
  code: string,
  framework: string,
  baseParse: (code: string) => any,
  parse: ParseFn,
  build: BuildFn,
  isTargetNode: IsTargetFn,
): T {
  const ast = baseParse(code)

  let replacedSource = code
  const sources: NodeInfo[] = []

  walk(ast as any, {
    enter(node: any) {
      if (isTargetNode(node)) {
        if (framework === 'eslint-typescript-parser') {
          replacedSource =
            replacedSource.slice(0, node.start) +
            ' '.repeat(node.argument.start - node.start) +
            replacedSource.slice(node.argument.start, node.argument.end) +
            ' '.repeat(node.end - node.argument.end) +
            replacedSource.slice(node.end)
          return
        }

        const valid = code.slice(node.argument.start, node.argument.end)
        sources.push([
          valid,
          node.start,
          node.end,
          node.argument.start,
          node.argument.end,
        ])

        const originalLength = node.end - node.start
        const placeholder = 'x'.repeat(originalLength)

        replacedSource = `${replacedSource.slice(0, node.start)}${placeholder}${replacedSource.slice(node.end)}`
      }
    },
  })

  const finalAST = parse(replacedSource, false)
  if (framework === 'eslint-typescript-parser') {
    return finalAST
  }

  let replaced = 0
  walk(finalAST as any, {
    enter(node, parent, key) {
      if (key === 'tokens') return

      if (['File', 'Program', 'ExpressionStatement'].includes(node.type)) {
        return
      }

      const [start, end] = getNodeRange(node)
      const valid = sources.find(
        ([, _start, _end]) => start === _start && end === _end,
      )
      if (!valid) return

      const newNode = build(start, end, processValid(valid))
      this.replace(newNode)
      this.skip()
    },
  })

  if (replaced !== sources.length) {
    throw new Error(
      `Expected to replace ${sources.length} nodes, but replaced ${replaced}.`,
    )
  }

  return finalAST

  function processValid(node: NodeInfo): any {
    replaced++
    const [, , , validStart, validEnd] = node

    const valid = sources.find(
      ([, start, end]) => start >= validStart && end <= validEnd,
    )
    if (valid) {
      const validNode = processValid(valid)
      const fullMatch = valid[1] === validStart && valid[2] === validEnd

      if (fullMatch) {
        return build(valid[1], valid[2], validNode)
      } else {
        throw new Error('Unsupported syntax')
      }
    } else {
      return parse(node[0], true, validStart)
    }
  }
}
