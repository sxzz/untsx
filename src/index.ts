import {
  parse as babelParse,
  parseExpression as babelParseExpression,
  type ParserOptions as BabelOptions,
  type ParseResult,
} from '@babel/parser'
import {
  parse as tsEslintParse,
  type TSESTree,
  type TSESTreeOptions,
} from '@typescript-eslint/typescript-estree'
import {
  Parser,
  type Options as AcornOptions,
  type Program as AcornProgram,
} from 'acorn'
import acornJsx from 'acorn-jsx'
import acornTS from 'acorn-typescript'
import { parse as espreeParse, type Options as EspreeOptions } from 'espree'
import { walk } from 'estree-walker'
import { createPrettierPlugin } from './prettier'
import { replace } from './replace'
import type { File } from '@babel/types'
import type { MagicStringAST } from 'magic-string-ast'
import type { Plugin as PrettierPlugin, Printer } from 'prettier'
import type { Codes } from 'ts-macro'

export type IsTargetFn = (node: any) => boolean
export type CustomAcornParser = (
  parser: typeof Parser,
) => typeof Parser | null | undefined

export const REGEX_TS: RegExp = /\.[cm]?tsx?$/
export const REGEX_JSX: RegExp = /\.[jt]sx$/

export interface UntsxFactory<T = any> {
  baseParser:
    | {
        name: 'acorn'
        customParser?: CustomAcornParser
      }
    | {
        name: 'babel'
        parserOptions: BabelOptions
      }
  isTarget: IsTargetFn
  build: (parserName: string, start: number, end: number, valid: any) => T
  shouldTransform?: (code: string, id: string) => boolean
  transform: (
    code: string,
    id: string,
    s: MagicStringAST | Codes,
    node: T,
    parent: any,
    key: PropertyKey | null | undefined,
    index: number | null | undefined,
    isVolar: boolean,
  ) => boolean
  format: Printer['print']
  getVisitorKeys?: Printer['getVisitorKeys']
}

export interface UntsxInstance {
  baseParse: (code: string, isTS?: boolean, isJSX?: boolean) => any
  acorn: (
    code: string,
    options: AcornOptions,
    isTS?: boolean,
    isJSX?: boolean,
    parser?: typeof Parser,
  ) => AcornProgram
  babel: (code: string, options?: BabelOptions) => ParseResult<File>
  espree: (code: string, options?: EspreeOptions) => AcornProgram
  eslintTypescript: (
    code: string,
    options?: TSESTreeOptions,
  ) => TSESTree.Program

  transform: (code: string, id: string, s: MagicStringAST | Codes) => boolean
  prettier: PrettierPlugin
}

export function createUntsx(factory: UntsxFactory): UntsxInstance {
  function baseParse(code: string, isTS?: boolean, isJSX?: boolean): any {
    if (factory.baseParser.name === 'acorn') {
      const parser = buildAcornParser(
        Parser,
        factory.baseParser.customParser,
        isTS,
        isJSX,
      )
      const ast = parser.parse(code, {
        ecmaVersion: 'latest',
        sourceType: 'module',
        allowAwaitOutsideFunction: true,
        allowHashBang: true,
        allowImportExportEverywhere: true,
        allowReserved: true,
        allowReturnOutsideFunction: true,
        allowSuperOutsideMethod: true,
      })
      return ast
    }

    return babelParse(code, {
      sourceType: 'module',
      errorRecovery: true,

      allowAwaitOutsideFunction: true,
      allowImportExportEverywhere: true,
      allowNewTargetOutsideFunction: true,
      allowReturnOutsideFunction: true,
      allowSuperOutsideMethod: true,
      allowUndeclaredExports: true,
      allowYieldOutsideFunction: true,

      ...factory.baseParser.parserOptions,
      plugins: [
        ...(isTS ? (['typescript'] as const) : []),
        ...(isJSX ? (['jsx'] as const) : []),
        ...(factory.baseParser.parserOptions.plugins || []),
      ],
    }).program
  }

  function acorn(
    code: string,
    options: AcornOptions,
    isTS?: boolean,
    isJSX?: boolean,
    parser: typeof Parser = Parser,
  ): AcornProgram {
    if (factory.baseParser.name === 'acorn') {
      parser = buildAcornParser(
        parser,
        factory.baseParser.customParser,
        isTS,
        isJSX,
      )
      return parser.parse(code, options)
    }

    return replace<AcornProgram>(
      code,
      'acorn',
      (code) => baseParse(code, isTS, isJSX),
      (code, isExpression) =>
        isExpression
          ? parser.parseExpressionAt(code, 0, options)
          : parser.parse(code, options),
      factory.build.bind(null, 'acorn'),
      factory.isTarget,
    )
  }

  function babel(code: string, options?: BabelOptions): ParseResult<File> {
    if (factory.baseParser.name === 'babel') {
      const { parserOptions } = factory.baseParser
      return babelParse(code, {
        ...parserOptions,
        ...options,
        plugins: [
          ...(parserOptions.plugins || []),
          ...(options?.plugins || []),
        ],
      })
    }

    function hasPlugin(plugin: string): boolean {
      if (!options?.plugins) return false
      return options.plugins.some((p) => p === plugin || p[0] === plugin)
    }

    return replace<ParseResult<File>>(
      code,
      'babel',
      (code) => baseParse(code, hasPlugin('typescript'), hasPlugin('jsx')),
      (code, isExpression) =>
        (isExpression ? babelParseExpression : babelParse)(code, options),
      factory.build.bind(null, 'babel'),
      factory.isTarget,
    )
  }

  function espree(code: string, options?: EspreeOptions): AcornProgram {
    return replace<AcornProgram>(
      code,
      'espree',
      () => baseParse(code, false, !!options?.ecmaFeatures?.jsx),
      (code, isExpression) => {
        const ast = espreeParse(isExpression ? `(${code})` : code, options)
        if (isExpression) {
          // @ts-expect-error
          return ast.body[0].expression
        }
        return ast
      },
      factory.build.bind(null, 'espree'),
      factory.isTarget,
    )
  }

  function eslintTypescript(code: string, options?: TSESTreeOptions) {
    return replace<TSESTree.Program>(
      code,
      'eslint-typescript',
      (code) => baseParse(code, true, !!options?.jsx),
      (code, isExpression) => {
        let ast = tsEslintParse(isExpression ? `(${code})` : code, {
          ...options,
          range: true,
        })

        // @ts-expect-error
        if (isExpression) ast = ast.body[0].expression

        return ast
      },
      // buildTryOperator,
      factory.build.bind(null, 'eslint-typescript'),
      factory.isTarget,
    )
  }

  function transform(
    code: string,
    id: string,
    s: MagicStringAST | Codes,
  ): boolean {
    const isVolar = Array.isArray(s)
    const ast = baseParse(code, REGEX_TS.test(id), id.endsWith('.jsx'))

    let changed = false
    walk(ast as any, {
      enter(
        node: any,
        parent: any,
        key: PropertyKey | null | undefined,
        index: number | null | undefined,
      ) {
        if (factory.isTarget(node)) {
          changed ||= factory.transform(
            code,
            id,
            s,
            node,
            parent,
            key,
            index,
            isVolar,
          )
        }
      },
    })

    return changed
  }

  const prettier = createPrettierPlugin(
    factory,
    acorn,
    babel,
    espree,
    eslintTypescript,
  )

  return {
    baseParse,
    acorn,
    babel,
    espree,
    eslintTypescript,
    transform,
    prettier,
  }
}

function buildAcornParser(
  parser: typeof Parser,
  customParser?: CustomAcornParser,
  isTS?: boolean,
  isJSX?: boolean,
) {
  if (isTS) {
    parser = parser.extend(acornTS() as any)
  } else if (isJSX) {
    parser = parser.extend(acornJsx())
  }
  parser = customParser?.(Parser) || parser
  return parser
}
