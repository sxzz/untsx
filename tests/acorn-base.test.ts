import path from 'node:path'
import { createTsMacroProgram } from '@sxzz/test-utils'
import { ESLint } from 'eslint'
import { MagicStringAST } from 'magic-string-ast'
import { format } from 'prettier'
import { tryOperatorPlugin } from 'tc39-try/acorn'
import { createPlugin } from 'ts-macro'
import { expect, test } from 'vitest'
import { createUntsx } from '../src'
import fixture from './fixtures/try?url'

const { acorn, babel, espree, eslintTypescript, transform, prettier, eslint } =
  createUntsx({
    baseParser: {
      name: 'acorn',
      customParser(parser) {
        return parser.extend(tryOperatorPlugin())
      },
    },
    isTarget: (node) =>
      node.type === 'UnaryExpression' && node.operator === 'try',
    build(parserName, start, end, valid) {
      return {
        type: 'UnaryExpression',
        operator: 'try',
        prefix: true,
        argument: valid,

        start,
        end,
        range: [start, end],
      }
    },
    transform(code, id, s, node) {
      const argumentStart = node.argument.start

      s.replaceRange(node.start, argumentStart, `() => (`)
      s.replaceRange(node.argument.end, node.end, ')')

      return true
    },
    format(path, options, print) {
      return ['try ', path.call(print, 'argument')]
    },
  })

const code = 'const x = try fn()'
const codeTS = 'const x: any = try fn()'

test('acorn', () => {
  const program = acorn(code, { ecmaVersion: 'latest' })
  const declaration = (program as any).body[0].declarations[0]

  expect(declaration.init).toMatchObject({
    type: 'UnaryExpression',
    operator: 'try',
    prefix: true,
    argument: {
      type: 'CallExpression',
      callee: {
        type: 'Identifier',
        name: 'fn',
        start: 14,
        end: 16,
      },
      arguments: [],
    },
    start: 10,
    end: 18,
  })
})

test('babel', () => {
  const program = babel(code)
  const declaration = (program as any).program.body[0].declarations[0]

  expect(declaration.init).toMatchObject({
    type: 'UnaryExpression',
    operator: 'try',
    prefix: true,
    argument: {
      type: 'CallExpression',
      callee: {
        type: 'Identifier',
        name: 'fn',
        start: 14,
        end: 16,
      },
      arguments: [],
    },
    start: 10,
    end: 18,
  })
})

test('babel with options', () => {
  const program = babel(codeTS, { ranges: true, plugins: ['typescript'] })
  const declaration = (program as any).program.body[0].declarations[0]

  expect(declaration.id).toMatchObject({
    type: 'Identifier',
    name: 'x',
    start: 6,
    end: 12,
    range: [6, 12],
    typeAnnotation: {
      type: 'TSTypeAnnotation',
      typeAnnotation: {
        type: 'TSAnyKeyword',
        start: 9,
        end: 12,
        range: [9, 12],
      },
      start: 7,
      end: 12,
      range: [7, 12],
    },
  })
  expect(declaration.init).toMatchObject({
    type: 'UnaryExpression',
    operator: 'try',
    prefix: true,
    argument: {
      type: 'CallExpression',
      callee: {
        type: 'Identifier',
        name: 'fn',
        start: 19,
        end: 21,
        range: [19, 21],
      },
      start: 19,
      end: 23,
      range: [19, 23],
    },
    start: 15,
    end: 23,
    range: [15, 23],
  })
})

test('espree', () => {
  const program = espree(code, { ecmaVersion: 'latest' })
  const declaration = (program as any).body[0].declarations[0]

  expect(declaration.init).toMatchObject({
    type: 'UnaryExpression',
    operator: 'try',
    prefix: true,
    argument: {
      type: 'CallExpression',
      callee: {
        type: 'Identifier',
        name: 'fn',
        start: 14,
        end: 16,
      },
      arguments: [],
    },
    start: 10,
    end: 18,
  })
})

test('typescript-eslint', () => {
  const program = eslintTypescript(code)
  const declaration = (program as any).body[0].declarations[0]

  expect(declaration.init).toMatchObject({
    type: 'UnaryExpression',
    operator: 'try',
    prefix: true,
    argument: {
      type: 'CallExpression',
      callee: {
        type: 'Identifier',
        name: 'fn',
        range: [14, 16],
      },
      arguments: [],
    },
    range: [10, 18],
  })
})

test('magic-string', () => {
  let s = new MagicStringAST(code)
  transform(s.toString(), 'test.js', s)
  expect(s.toString()).toBe('const x = () => (fn())')

  s = new MagicStringAST(codeTS)
  transform(s.toString(), 'test.ts', s)
  expect(s.toString()).toBe('const x: any = () => (fn())')
})

test('volar', () => {
  const plugin = createPlugin(() => ({
    name: 'test-plugin',
    resolveVirtualCode({ codes, filePath, ast }) {
      transform(ast.text, filePath, codes)
    },
  }))

  const program = createTsMacroProgram([path.resolve(`.${fixture}`)], [plugin])
  const files = program
    .getSourceFiles()
    .filter((file) => !file.fileName.includes('node_modules'))
  for (const file of files) {
    expect(file.text.trim()).matchSnapshot()
  }
})

test.each(['acorn', 'babel', 'babel-ts', 'typescript'])(
  'prettier: parser %s',
  async (parser) => {
    const isTS = parser === 'babel-ts' || parser === 'typescript'
    const code = isTS ? 'const  x:never  =  try  1;' : 'const  x  =  try 1;'
    const result = await format(code, {
      parser,
      plugins: [prettier],
      semi: false,
    })
    const formatted = isTS ? 'const x: never = try 1' : 'const x = try 1'
    expect(result).toBe(`${formatted}\n`)
  },
)

test('eslint', async () => {
  const linter = new ESLint({
    overrideConfigFile: true,
    baseConfig: [
      {
        files: ['**/*.js'],
        languageOptions: { parser: eslint.jsParser },
      },
      {
        files: ['**/*.ts'],
        languageOptions: { parser: eslint.tsParser },
      },
    ],
  })
  {
    const results = await linter.lintText(code, { filePath: 'test.js' })
    expect(results[0].messages).toHaveLength(0)
  }
  {
    const results = await linter.lintText(codeTS, { filePath: 'test.ts' })
    expect(results[0].messages).toHaveLength(0)
  }
})
