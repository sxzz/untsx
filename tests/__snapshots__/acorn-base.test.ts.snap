// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`volar 1`] = `"export const x = try fn()"`;
